rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -1067,15 +1185,82 @@ impl Program {
                 let inspect = with_prof_context(&description, || {
                     col.inspect(move |(v, ts, w)| ifun(v, ts.to_tuple_ts(), *w))
                 });
-                Self::xform_collection(inspect, &*next, arrangements)
+                Self::xform_collection(inspect, &*next, arrangements, lookup_collection)
+            }
+            XFormCollection::StreamJoin {
+                ref description,
+                afun,
+                arrangement,
+                jfun,
+                ref next,
+            } => {
+                let join = with_prof_context(&description, || {
+                    // arrange input collection
+                    let collection_with_keys = col.flat_map(afun);
+                    let arr = match arrangements.lookup_arr(arrangement) {
+                        A::Arrangement1(ArrangedCollection::Map(arranged)) => arranged.clone(),
+                        A::Arrangement1(ArrangedCollection::Set(_)) => {
+                            panic!("StreamJoin: not a map arrangement {:?}", arrangement)
+                        }
+                        _ => panic!("StreamJoin in nested scope: {}", description),
+                    };
+                    lookup_map(
+                        &collection_with_keys,
+                        arr,
+                        |(k, _), key| *key = k.clone(),
+                        move |v1, w1, v2, w2| (jfun(&v1.1, v2), w1 * w2),
+                        ().into_ddvalue(),
+                        ().into_ddvalue(),
+                        ().into_ddvalue(),
+                    )
+                    // Filter out `None`'s.
+                    // FIXME: We wouldn't need this if `lookup_map` allowed `output_func`
+                    // to return `Option`.
+                    .flat_map(|v| v)
```

Comment:
Nit, I think using the `identity` function is clearer but it's a matter of personal preference